gsplat/cuda/csrc/rasterization.cu

```diff
@@ -0,0 +1,1245 @@
+
+            T = next_T;
+        }
+    }
+
+    if (inside) {
+        // Here T is the transmittance AFTER the last gaussian in this pixel.
+        // We (should) store double precision as T would be used in backward pass and
+        // it can be very small and causing large diff in gradients with float32.
+        // However, double precision makes the backward pass 1.5x slower so we stick
+        // with float for now.
+        render_alphas[pix_id] = 1.0f - T;
+        PRAGMA_UNROLL
+        for (int k = 0; k < COLOR_DIM; ++k) {
+            render_colors[pix_id * COLOR_DIM + k] =
+                backgrounds == nullptr ? pix_out[k] : (pix_out[k] + T * backgrounds[k]);
+        }
+        // index in bin of last gaussian in this pixel
+        last_ids[pix_id] = cur_idx;
+    }
+}
+
+std::tuple<torch::Tensor, torch::Tensor, torch::Tensor> rasterize_to_pixels_fwd_tensor(
+    // Gaussian parameters
+    const torch::Tensor &means2d,   // [C, N, 2] or [nnz, 2]
+    const torch::Tensor &conics,    // [C, N, 3] or [nnz, 3]
+    const torch::Tensor &colors,    // [C, N, channels] or [nnz, channels]
+    const torch::Tensor &opacities, // [C, N]  or [nnz]
+    const at::optional<torch::Tensor> &backgrounds, // [C, channels]
+    // image size
+    const int image_width, const int image_height, const int tile_size,
+    // intersections
+    const torch::Tensor &tile_offsets, // [C, tile_height, tile_width]
+    const torch::Tensor &gauss_ids     // [n_isects]
+) {
+    DEVICE_GUARD(means2d);
+    CHECK_INPUT(means2d);
+    CHECK_INPUT(conics);
+    CHECK_INPUT(colors);
+    CHECK_INPUT(opacities);
+    CHECK_INPUT(tile_offsets);
+    CHECK_INPUT(gauss_ids);
+    if (backgrounds.has_value()) {
+        CHECK_INPUT(backgrounds.value());
+    }
+    bool packed = means2d.dim() == 2;
+
+    int C = tile_offsets.size(0);          // number of cameras
+    int N = packed ? -1 : means2d.size(1); // number of gaussians
+    int channels = colors.size(-1);
+    int tile_height = tile_offsets.size(1);
+    int tile_width = tile_offsets.size(2);
+    int n_isects = gauss_ids.size(0);
+
+    // Each block covers a tile on the image. In total there are
+    // C * tile_height * tile_width blocks.
+    dim3 threads = {tile_size, tile_size, 1};
+    dim3 blocks = {C, tile_height, tile_width};
+
+    torch::Tensor renders = torch::empty({C, image_height, image_width, channels},
+                                         means2d.options().dtype(torch::kFloat32));
+    torch::Tensor alphas = torch::empty({C, image_height, image_width, 1},
+                                        means2d.options().dtype(torch::kFloat32));
+    torch::Tensor last_ids = torch::empty({C, image_height, image_width},
+                                          means2d.options().dtype(torch::kInt32));
+
+    at::cuda::CUDAStream stream = at::cuda::getCurrentCUDAStream();
+
+    // TODO: an optimization can be done by passing the actual number of channels into
```

Comment:
good idea, though not necessary for this version unless you're motivated XD

Reply:
I'll leave this to next PR ... lol